File: rust/bob/tests/bob.rs
use bob::*;

#[test]
fn asking_a_question() {
	let remark = "How are you?";
	assert_eq!(reply(remark), "Sure.");
}

#[test]
fn yelling() {
	let remark = "WHAT THE HECK";
	assert_eq!(reply(remark), "Whoa, chill out!");
}

#[test]
fn yelling_question() {
	let remark = "WHAT THE HECK?";
	assert_eq!(reply(remark), "Calm down, I know what I'm doing!");
}

#[test]
fn silence() {
	let remark = "    \t";
	assert_eq!(reply(remark), "Fine. Be that way!");
}

#[test]
fn regular() {
	let remark = "What the heck";
	assert_eq!(reply(remark), "Whatever.");
}

#[test]
fn non_letters_with_question() {
	let remark = ":) ?";
	assert_eq!(reply(remark), "Sure.");
}

#[test]
fn no_letters() {
	let remark = "1, 2, 3";
	assert_eq!(reply(remark), "Whatever.");
}

#[test]
fn question_with_no_letters() {
	let remark = "4?";
	assert_eq!(reply(remark), "Sure.");
}

File: typescript/bob/bob.ts
// function isAllUpperCase(text: string): boolean {
//   for (let i = 0; i < text.length; i++) {
//     const char = text[i];
//     if (char >= 'A' && char <= 'Z') {
//       return true;
//     }
//   }
//   return false;
// }



export function hey(message: string): string {
  const hasLetter = /[a-zA-Z]/.test(message); //包含字母
  const upperCase = message.toUpperCase(); 
  const lastText = message.trim().at(-1);
  if(hasLetter && (upperCase === message) && (lastText === "?")){
    return "Calm down, I know what I'm doing!"
  } //包含字母并且全大写并以问号结尾的情况。
  if(message.trim().at(-1) === "?"){
    return "Sure."
  } // 选出清空 string 里的空格后，最后一位是问号。
  if(message.trim() === ""){
    return "Fine. Be that way!"
  }  //选出只留空格的。 
  if(hasLetter && (upperCase === message)){
    return "Whoa, chill out!"
  } //如果本身=大写后的本身，说明本身全是大写。
  return "Whatever."
}
